File: DP(Dynamic_Programming, 동적계획법)/백준_2579_계단오르기/계단오르기.py
stairs = [0 for _ in range(302)]
sum = [0 for _ in range(302)]

n = int(input())
for i in range(1, n+1):
    stairs[i] = int(input())
sum[1] = stairs[1]
sum[2] = stairs[1] + stairs[2]
sum[3] = max(stairs[1], stairs[2]) + stairs[3]
for i in range(4, n+1):
    sum[i] = max(sum[i-3] + stairs[i-1], sum[i-2]) + stairs[i]

print(sum[n])

File: DP(Dynamic_Programming, 동적계획법)/백준_2579_계단오르기/계단오르기.cs
using System.Text;
using System;
using static System.Console;

class Program
{
    static int[] stairs = new int[302];
    static int[] sum = new int[302];
    public static void Main(string[] args)
    {
        int n = int.Parse(ReadLine());
        for(int i =1; i <= n; i++)
        {
            stairs[i] = int.Parse(ReadLine());
        }

        sum[1] = stairs[1];
        sum[2] = stairs[1] + stairs[2];
        sum[3] = Math.Max(stairs[1], stairs[2]) + stairs[3];

        for(int i = 4; i <= n; i++)
        {
            sum[i] = Math.Max(sum[i - 3] + stairs[i - 1], sum[i - 2]) + stairs[i];
        }
        Write(sum[n]);
    }
}